lib/src/core/engine.dart

```diff
@@ -599,21 +549,137 @@ class Engine extends Disposable with EventsEmittable<EngineEvent> {
     }
   }
 
-  Future<void> _onDisconnected(DisconnectReason reason) async {
+  Future<void> handleDisconnect(DisconnectReason reason) async {
     logger
         .info('onDisconnected state:${_connectionState} reason:${reason.name}');
+
+    if (!fullReconnect) {
+      fullReconnect = _clientConfiguration?.resumeConnection ==
+              lk_models.ClientConfigSetting.DISABLED ||
+          [
+            DisconnectReason.leaveReconnect,
+            DisconnectReason.negotiationFailed,
+            DisconnectReason.peerConnectionClosed
+          ].contains(reason);
+    }
+
+    if (_connectionState == ConnectionState.reconnecting && !fullReconnect) {
+      logger.fine('[$objectId] Already reconnecting...');
+      return;
+    }
+
     if (_connectionState == ConnectionState.disconnected) {
       logger.fine('[$objectId] Already disconnected... $reason');
       return;
     }
-    if (_connectionState == ConnectionState.reconnecting) {
-      logger.fine('[$objectId] Already reconnecting...');
+
+    logger.fine('[$runtimeType] Should attempt reconnect sequence...');
+    if (fullReconnect) {
+      await restartConnection();
+    } else {
+      bool reConnectSignal = [
+        DisconnectReason.leaveReconnect,
+        DisconnectReason.peerConnectionClosed,
+        DisconnectReason.signal
+      ].contains(reason);
+      await resumeConnection(reConnectSignal);
+    }
+  }
+
+  Future<void> resumeConnection([bool reConnectSignal = false]) async {
```

Comment:
this seems incorrect as well. in JS, even with a resume, it should *always* reconnect to signal. The difference is:
* `reconnect: false` when it's a full reconnect
* `reconnect: true` when it's a resume (ICE restart or migration)

we always reconnect signal in case the old server has failed.

Reply:
in JS, we take in a parameter in order to know whether or not to emit a `Resuming` event.

Reply:
My intention to add reConnectSignal here is to trigger an iceRestart to reconnect the peerconnection when the signaling is connected but the ice connection is temporarily disconnected.

Reply:
we are not able to distinguish peerconnection temporarily disconnected vs the server is gone. Because of that, we should always reconnect the signal connection because it will start performing a media server discovery.

Reply:
okay, makes sense

Reply:
resolved https://github.com/livekit/client-sdk-flutter/pull/194/commits/678c430949d559ce6b1bcbda525a6a74fedcbe13